src/main/java/appeng/core/features/registries/GridCacheRegistry.java

```diff
@@ -18,54 +18,57 @@
 
 package appeng.core.features.registries;
 
-import java.lang.reflect.Constructor;
-import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
+import javax.annotation.Nonnull;
+
 import appeng.api.networking.IGrid;
 import appeng.api.networking.IGridCache;
+import appeng.api.networking.IGridCacheFactory;
 import appeng.api.networking.IGridCacheRegistry;
 import appeng.core.AELog;
 
 public final class GridCacheRegistry implements IGridCacheRegistry {
-    private final Map<Class<? extends IGridCache>, Class<? extends IGridCache>> caches = new HashMap<>();
+
+    private final List<GridCacheRegistration<?>> registry = new ArrayList<>();
 
     @Override
-    public void registerGridCache(final Class<? extends IGridCache> iface,
-            final Class<? extends IGridCache> implementation) {
-        if (iface.isAssignableFrom(implementation)) {
-            this.caches.put(iface, implementation);
-        } else {
-            throw new IllegalArgumentException(
-                    "Invalid setup, grid cache must either be the same class, or an interface that the implementation implements. Gotten: "
-                            + iface + " and " + implementation);
+    public synchronized <T extends IGridCache> void registerGridCache(@Nonnull Class<T> iface,
+            @Nonnull IGridCacheFactory<T> factory) {
+
+        if (registry.stream().anyMatch(r -> r.cacheClass.equals(iface))) {
+            AELog.debug("Overriding grid cache factory for cache type %s", iface);
```

Comment:
This should really crash not just log some debug statement. Replacing core behaviour should never be possible for addons.

Reply:
Before this change, it just silently swallowed it :-D
So we're in agreement, I'll push the change.